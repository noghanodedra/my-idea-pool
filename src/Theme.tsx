import React, { ReactNode } from 'react';
import { ThemeProvider } from 'styled-components';

// Define the `theme` prop

export const theme = {
  main: {
    color: {
      primary: "#00A843",
      inputBorder: "#455E70",
      placeholder: "#2a3842",
    },
  },
};

interface IProps {
  children: ReactNode;
}

const Theme = ({ children }: IProps) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
);

export default Theme;
