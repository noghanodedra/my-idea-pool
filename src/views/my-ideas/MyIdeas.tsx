import { UserContext } from 'contexts/UserContext';
import React, { useContext, useEffect, useState } from 'react';
import InfiniteScroll from 'react-infinite-scroller';
import { ideaService } from 'services/idea-service';
import { userService } from 'services/user-service';
import styled from 'styled-components';

import AddIcon from '../../assets/images/btn_addanidea@2x.png';
import BulbIcon from '../../assets/images/bulb@2x.png';
import { Idea } from '../../models/Idea';
import Header from './components/header/Header';
import InlineEditRow from './components/inline-edit-row/InlineEditRow';

const StyledContainer = styled.div`
  margin-left: 79px;
  margin-right: 87px;
  height: 100%;
  width: 100%;
`;

const StyledHeader = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  border-bottom: 1px solid rgba(42, 56, 66, 0.2);
`;

const StyledTitle = styled.div`
  font-size: 28px;
  color: #2a3842;
  margin: 41px 0 47px 20px;
`;

const AddButtonImg = styled.img`
  width: 50px;
  height: 50px;
  cursor: pointer;
  padding-right: 10px;
`;

const StyledIdeasContainer = styled.div`
  text-align: center;
  height: 750px;
  overflow-x: hidden;
  overflow-y: auto;
`;

const StyledNoRecordsView = styled.div`
  font-size: 20px;
  color: #2a3842;
  display: flex;
  justify-content: center;
  align-items: center;
  align-content: center;
  flex-direction: column;
`;

const BulbImg = styled.img`
  width: 64px;
  height: 96px;
  margin-top: 246px;
  margin-bottom: 23px;
`;

const defaultRecord: Idea = {
  id: "",
  content: "",
  impact: 10,
  ease: 9,
  confidence: 8,
};

const MyIdeas = () => {
  const initialRecords: Idea[] = [];
  const [records, setRecords] = useState(initialRecords);
  const [addNewRecord, setAddNewRecord] = useState(false);
  const [loading, setLoading] = useState(false);

  let pageNumber = 1;

  const { setDetails } = useContext(UserContext);

  const getProfileDetails = async () => {
    try {
      const response = await userService.me();
      setDetails(response.data);
      return response.data;
    } catch (e) {
    }
    return null;
  };

  const getIdeas = async (pageNumber: number = 1) => {
    try {
      setLoading(true);
      const response = await ideaService.getIdeas(pageNumber);
      setRecords([]);
      setRecords(response.data);
      setLoading(false);
      return response.data;
    } catch (e) {
    }
    return null;
  };

  const onAdd = () => {
    setAddNewRecord(true);
  };

  const loadFunc = () => {
    pageNumber++;
    getIdeas(pageNumber);
  };

  useEffect(() => {
    getProfileDetails();
    getIdeas(pageNumber);
    return () => {};
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  let scrollParentRef: any = null;

  return (
    <StyledContainer>
      <StyledHeader>
        <StyledTitle>My Ideas</StyledTitle>
        <AddButtonImg src={AddIcon} onClick={onAdd} alt="addRecord"></AddButtonImg>
      </StyledHeader>
      <StyledIdeasContainer ref={(ref) => (scrollParentRef = ref)}>
        {records.length === 0 && !addNewRecord && !loading ? (
          <StyledNoRecordsView>
            <BulbImg src={BulbIcon}></BulbImg>
            Got Ideas?
          </StyledNoRecordsView>
        ) : (
          <InfiniteScroll
            pageStart={0}
            loadMore={loadFunc}
            hasMore={false}
            useWindow={false}
            getScrollParent={() => scrollParentRef}
            loader={
              <div className="loader" key={0}>
                Loading ...
              </div>
            }
          >
            <table>
              <thead>
                <Header></Header>
              </thead>
              <tbody>
                {addNewRecord && (
                  <InlineEditRow
                    key={1}
                    record={defaultRecord}
                    removeInlineEditFn={setAddNewRecord}
                    recordsLoaderFn={getIdeas}
                  ></InlineEditRow>
                )}
                {records.map((record, index) => (
                  <InlineEditRow
                    key={record.id}
                    record={record}
                    recordsLoaderFn={getIdeas}
                  ></InlineEditRow>
                ))}
              </tbody>
            </table>
          </InfiniteScroll>
        )}
      </StyledIdeasContainer>
    </StyledContainer>
  );
};

export default MyIdeas;
