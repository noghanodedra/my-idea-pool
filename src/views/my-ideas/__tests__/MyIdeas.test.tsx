import 'jest-styled-components';
import 'mutationobserver-shim';

import { fireEvent, render, screen } from '@testing-library/react';
import { shallow } from 'enzyme';
import React from 'react';
import Theme from 'Theme';

import MyIdeas from '../MyIdeas';


const renderComponent = () =>
  render(
    <Theme>
      <MyIdeas></MyIdeas>
    </Theme>
  );

describe("MyIdeas Component", () => {
  beforeEach(() => {
    renderComponent();
  });

  it("renders correctly", async () => {
    expect(await screen.findAllByText(/My Ideas/i)).toHaveLength(1);
    expect(await screen.findAllByText(/Got Ideas/i)).toHaveLength(1);
     expect(
       await screen.getAllByText(
         (content, element) => element.tagName.toLowerCase() === "img"
       )
     ).toHaveLength(1);
  });

  it("is styled correctly", () => {
    const wrapper = shallow(
      <Theme>
        <MyIdeas></MyIdeas>
      </Theme>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should call onAdd on plus image click", async () => {

    fireEvent.click(screen.getByAltText("addRecord"));

    expect(await screen.findAllByText(/Impact/i)).toHaveLength(1);
    expect(
      await screen.getAllByText(
        (content, element) => element.tagName.toLowerCase() === "img"
      )
    ).toHaveLength(3);
  });

});
