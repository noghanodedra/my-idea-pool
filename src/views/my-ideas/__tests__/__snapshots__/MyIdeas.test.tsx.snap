// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyIdeas Component is styled correctly 1`] = `
<ThemeProvider
  theme={
    Object {
      "main": Object {
        "color": Object {
          "inputBorder": "#455E70",
          "placeholder": "#2a3842",
          "primary": "#00A843",
        },
      },
    }
  }
>
  <MyIdeas />
</ThemeProvider>
`;

exports[`Numeric Stepper Component is styled correctly 1`] = `
<ThemeProvider
  theme={
    Object {
      "main": Object {
        "color": Object {
          "inputBorder": "#455E70",
          "placeholder": "#2a3842",
          "primary": "#00A843",
        },
      },
    }
  }
>
  <MyIdeas />
</ThemeProvider>
`;
